pkg/cloud/vsphere/provisioner/govmomi/delete.go

```diff
@@ -20,52 +21,69 @@ func (pv *Provisioner) Delete(ctx context.Context, cluster *clusterv1.Cluster, m
 	if err != nil {
 		return err
 	}
 	deletectx, cancel := context.WithCancel(*s.context)
 	defer cancel()
 
-	if exists, _ := pv.Exists(ctx, cluster, machine); exists {
-		moref, err := vsphereutils.GetMachineRef(machine)
+	moref, err := vsphereutils.GetMachineRef(machine)
+	if err != nil {
+		return err
+	}
+	var vm mo.VirtualMachine
+	vmref := types.ManagedObjectReference{
+		Type:  "VirtualMachine",
+		Value: moref,
+	}
+	err = s.session.RetrieveOne(deletectx, vmref, []string{"name", "runtime.powerState"}, &vm)
+	if err != nil {
+		return err
+	}
+	pv.eventRecorder.Eventf(machine, corev1.EventTypeNormal, "Killing", "Killing machine %v", machine.Name)
+	vmo := object.NewVirtualMachine(s.session.Client, vmref)
+	if vm.Runtime.PowerState == types.VirtualMachinePowerStatePoweredOn {
+		task, err := vmo.PowerOff(deletectx)
 		if err != nil {
+			klog.Infof("Error trigerring power off operation on the Virtual Machine %s", vm.Name)
 			return err
 		}
-		var vm mo.VirtualMachine
-		vmref := types.ManagedObjectReference{
-			Type:  "VirtualMachine",
-			Value: moref,
-		}
-		err = s.session.RetrieveOne(deletectx, vmref, []string{"name", "runtime.powerState"}, &vm)
+		err = task.Wait(deletectx)
 		if err != nil {
+			klog.Infof("Error powering off the Virtual Machine %s", vm.Name)
 			return err
 		}
-		pv.eventRecorder.Eventf(machine, corev1.EventTypeNormal, "Killing", "Killing machine %v", machine.Name)
-		vmo := object.NewVirtualMachine(s.session.Client, vmref)
-		if vm.Runtime.PowerState == types.VirtualMachinePowerStatePoweredOn {
-			task, err := vmo.PowerOff(deletectx)
-			if err != nil {
-				klog.Infof("Error trigerring power off operation on the Virtual Machine %s", vm.Name)
-				return err
-			}
-			err = task.Wait(deletectx)
-			if err != nil {
-				klog.Infof("Error powering off the Virtual Machine %s", vm.Name)
-				return err
-			}
-		}
-		task, err := vmo.Destroy(deletectx)
-		taskinfo, err := task.WaitForResult(deletectx, nil)
-		if taskinfo.State == types.TaskInfoStateSuccess {
-			klog.Infof("Virtual Machine %v deleted successfully", vm.Name)
-			pv.eventRecorder.Eventf(machine, corev1.EventTypeNormal, "Killed", "Machine %v deletion complete", machine.Name)
-			return nil
-		}
+	}
+	task, err := vmo.Destroy(deletectx)
+	taskinfo, err := task.WaitForResult(deletectx, nil)
+	if taskinfo.State == types.TaskInfoStateSuccess {
+		klog.Infof("Virtual Machine %v deleted successfully", vm.Name)
 		pv.eventRecorder.Eventf(machine, corev1.EventTypeNormal, "Killed", "Machine %v deletion complete", machine.Name)
-		klog.Errorf("VM Deletion failed on pv with following reason %v", taskinfo.Reason)
-		return errors.New("VM Deletion failed")
+		return nil
 	}
-	return nil
+	pv.eventRecorder.Eventf(machine, corev1.EventTypeNormal, "Killed", "Machine %v deletion complete", machine.Name)
+	klog.Errorf("VM Deletion failed on pv with following reason %v", taskinfo.Reason)
+	return errors.New("VM Deletion failed")
```

Comment:
Should we return the name of the Virtual Machine as part of the error? Might help with debugging

Reply:
done